### PR TITLE
Reduxify: lib/shortcodes, take two

### PR DESCRIPTION
**Pull request by blowery**
The reapplication of #8857, with a slightly simpler implementation using the new action proxy.

To test:

* Checkout this branch
* Verify there are no regressions within:
  * Post editor shortcodes (rendering and editing items, captions, inserting, reordering and deleting items)
  * Post editor media modal (editing media item data, including title and secondary fields)
  * Post editor featured image (rendering and editing)
  * Media library - dropping a file, uploading a file through URL, uploading a file directly
* Verify all tests pass
* Verify nothing imports `lib/shortcodes`

### CODE REVIEW

File: client/components/gallery-shortcode/index.jsx [hunk 1]
```diff
@@ -83,7 +83,7 @@ export default class GalleryShortcode extends React.Component {
 	};
 
 	getAttributes = () => {
-		let attributes = pick( this.props, 'items', 'type', 'columns', 'orderBy', 'link', 'size' );
+		const attributes = pick( this.props, 'items', 'type', 'columns', 'orderBy', 'link', 'size' );
```
Comment (jsnajdr) [inline client/components/gallery-shortcode/index.jsx]: There's one possible improvement in the `GalleryShortcode` component that would be an awesome drive-by patch: rewrite this line:
```js
omit( this.props, Object.keys( this.constructor.propTypes ) )
```
to something that doesn't access the component's `propTypes` at runtime. If we can eliminate all code like this, we could remove all proptypes definitions from production bundles.

File: client/blocks/shortcode/index.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	const classes = classNames( 'shortcode', className );
+	let filteredShortcode = {};
+	if ( shortcode ) {
+		shortcode.body = shortcode.result;
+		filteredShortcode = filterRenderResult( omit( shortcode, 'shortcode' ) );
+	}
+	return (
+		<div>
+			<QueryShortcode siteId={ siteId } shortcode={ children } />
+			<ShortcodeFrame
```
Comment (jsnajdr) [inline client/blocks/shortcode/index.js]: After the Redux refactoring, is the `ShortcodeFrame` still needed as a standalone component? Everything important happens in `connect` now and the actual `Shortcode` component is a very thin wrapper around `ShortcodeFrame`.

File: client/blocks/shortcode/index.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+				{ ...omit( props, 'siteId', 'filterRenderResult', 'shortcode', 'dispatch' ) }
```
Comment (jsnajdr) [inline client/blocks/shortcode/index.js]: These `omit` calls make it very hard to figure out what the component's API actually is and what props are possible.

But they are used everywhere in the related components (`GalleryShortcode` et al) and it will be hard to get rid of them. Topic for a standalone PR rather than additional change in this one.

File: client/state/shortcodes/reducer.js [hunk 4]
```diff
@@ -46,6 +51,36 @@ export const requesting = createReducer(
...
 
+function mediaItemsReducer( state, { siteId, data } ) {
+	if ( ! state.hasOwnProperty( siteId ) ) {
+		return state;
+	}
+
+	if ( ! ( data && data.media ) ) {
+		return state;
+	}
+	const media = castArray( data.media );
```
Comment (jsnajdr) [inline client/state/shortcodes/reducer.js]: This will not work on a `RECEIVE_MEDIA_ITEM` dispatch. In this case, `data`'s value is the media item and there's no `data.media` property. At least that's how I understand the original Flux code at https://github.com/Automattic/wp-calypso/blob/master/client/lib/shortcodes/store.js#L73-L78

File: client/state/shortcodes/reducer.js [hunk 5]
```diff
@@ -46,6 +51,36 @@ export const requesting = createReducer(
...
+	const updatedIds = media.map( item => String( item.ID ) );
+
+	return {
+		...state,
+		[ siteId ]: pickBy( state[ siteId ], shortcode => {
```
Comment (jsnajdr) [inline client/state/shortcodes/reducer.js]: This reducer will always mutate the state even if no matching shortcode is found. To get an A grade, it should return unmodified state.

File: client/state/shortcodes/reducer.js [hunk 6]
```diff
@@ -17,6 +17,11 @@ import {
 	SHORTCODE_REQUEST_FAILURE,
 	SHORTCODE_REQUEST_SUCCESS,
 } from 'state/action-types';
+import { registerActionForward } from 'lib/redux-bridge';
+import { parse } from 'lib/shortcode';
+
+registerActionForward( 'RECEIVE_MEDIA_ITEMS' );
+registerActionForward( 'RECEIVE_MEDIA_ITEM' );
 
 const createRequestingReducer = requesting => {
```
Comment (jsnajdr) [inline client/state/shortcodes/reducer.js]: If the request was implemented with data layer, its deduplication feature could replace the `requesting` flag. Potential future improvement.

File: client/state/shortcodes/reducer.js (hunk 5 above)
Comment (jsnajdr) [reply client/state/shortcodes/reducer.js]: All reducers in this module would benefit from using `keyedReducer`. Potential future improvement.

**Review by jsnajdr — CHANGES_REQUESTED** [review]
I believe I found a bug when dispatching `RECEIVE_MEDIA_ITEM`. My other comments are not blockers, just suggestions.

File: client/state/shortcodes/reducer.js (hunk 4 above)
Comment (blowery) [reply client/state/shortcodes/reducer.js]: nice catch! I totally misread that line.

**Comment by jsnajdr** [comment]
The `Shortcode` component will be a very good candidate to try out React Suspense once it's released in some usable state. It uses the very same basic pattern (convert shortcode data into a markup asynchronously and cache the results) as all the demos.

File: client/blocks/shortcode/index.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ */
+import ShortcodeFrame from './frame';
+
+const Shortcode = props => {
+	const { siteId, className, children, filterRenderResult, shortcode } = props;
+	const classes = classNames( 'shortcode', className );
+	let filteredShortcode = {};
+	if ( shortcode ) {
+		shortcode.body = shortcode.result;
+		filteredShortcode = filterRenderResult( omit( shortcode, 'shortcode' ) );
```
[the remark this answers is not in the document]
Comment (blowery) [reply client/blocks/shortcode/index.js]: it could be, but it makes `mapStateToProps` more complicated. Not sure it's worth it.

File: client/blocks/shortcode/index.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	return (
+		<div>
```
[the remark this answers is not in the document]
Comment (blowery) [reply client/blocks/shortcode/index.js]: done

File: client/state/shortcodes/reducer.js (hunk 6 above)
Comment (blowery) [reply client/state/shortcodes/reducer.js]: totally

File: client/state/shortcodes/reducer.js (hunk 5 above)
Comment (blowery) [reply client/state/shortcodes/reducer.js]: yup, future pr

File: client/blocks/shortcode/index.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+		</Fragment>
+	);
+};
+
+Shortcode.propTypes = {
+	siteId: PropTypes.number.isRequired,
+	children: PropTypes.string.isRequired,
+	filterRenderResult: PropTypes.func,
+	className: PropTypes.string,
+	shortcode: PropTypes.string,
```
Comment (jsnajdr) [inline client/blocks/shortcode/index.js]: I'm getting a warning for the `shortcode` prop:
> Warning: Failed prop type: Invalid prop `shortcode` of type `object` supplied to `Shortcode`, expected `string`.

The value returned by the `getShortcode` selector is an object.

**Review by jsnajdr — APPROVED** [review]
There's a proptype warning on `Shortcode`'s `shortcode` prop. Other than that, the patch works very well.

File: client/blocks/shortcode/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jsnajdr) [reply client/blocks/shortcode/index.js]: As the `Shortcode` component doesn't have any default value for `filterRenderResult`, this line crashes when the prop is not specified. Like in the `VideoView`. Test with a post that has a `[wpvideo]` shortcode in its content.

The `ShortcodeData` component used to specify a default value (identity), but it was lost during the refactoring.

File: client/blocks/shortcode/index.js (hunk 9 above)
Comment (blowery) [reply client/blocks/shortcode/index.js]: bah.

Comment (blowery) [reply client/blocks/shortcode/index.js]: fixed
